server/serializers.py

```diff
@@ -68,7 +71,29 @@ def _camelize_value(v: Any) -> Any:
         return v
 
 
-JsonizeContext = namedtuple("JsonizeContext", ["user", "session", "locale_id"])
+JsonizeContext = namedtuple(
+    "JsonizeContext", ["user", "session", "locale_id", "localizers"]
+)
+JsonizeModuleContext = namedtuple(
+    "JsonizeModuleContext", ["user", "session", "locale_id", "module_id", "localizers"]
+)
+
+
+def _add_module_to_ctx(ctx: JsonizeContext, module_id: str) -> JsonizeModuleContext:
+    return JsonizeModuleContext(
+        user=ctx.user,
+        session=ctx.session,
+        locale_id=ctx.locale_id,
+        module_id=module_id,
+        localizers=ctx.localizers,
+    )
+
+
+def collect_i18n_message_sources(locale_id: str):
+    return {
+        f"module.{module_id}": MessageLocalizer.for_module(module_zip, locale_id)
+        for module_id, module_zip in MODULE_REGISTRY.all_latest().items()
```

Comment:
You cannot call `MODULE_REGISTRY.all_latest()` here, because it hits the database from the event-loop thread. I'll add a comment to `ModuleRegistry` warning against this pattern.

Nix `collect_i18n_message_sources()`. It is unusable. And we don't need to -- a cache in `MessageLocalizer` should be faster, anyway.

Reply:
The question then is who will call `MODULE_REGISTRY.all_latest()` (via something like `MESSAGE_LOCALIZER_REGISTRY.update_modules()`).

Since `MODULE_REGISTRY` actually changes in only a few circumtances, we could initialize `MESSAGE_LOCALIZER_REGISTRY` on system startup and only run `MESSAGE_LOCALIZER_REGISTRY.update_modules()` when doing stuff that changes modules. I tried doing this (among other stuff...) in commits https://github.com/CJWorkbench/cjworkbench/pull/176/commits/441eaf54cf4be4afe5119d03c7d9c36061064ac0 and https://github.com/CJWorkbench/cjworkbench/pull/176/commits/6a242d22446de1ca781b8bf925101d882655eeec; however, adding `MESSAGE_LOCALIZER_REGISTRY.update_modules()` to asgi creation broke the setting up of `server.tests.test_websockets`.

Reply:
I don't understand the problem. `MODULE_REGISTRY.all_latest()` is already called. The `clientside.Module` is already passed to `jsonize_clientside_update()`.

Reply:
Ah, I see the problem. Wanna chat about it on Slack?